src/lua/xml.c

```diff
@@ -0,0 +1,318 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/xml.h"
+
+#include <assert.h>
+#include <ctype.h>
+#include <lua.h>
+#include <lauxlib.h>
+#include <stddef.h>
+#include <string.h>
+
+#include "lua/utils.h"
+
+int
+luaT_xml_decode(struct lua_State *L)
+{
+	/* The input string must be at the top of the stack. */
+	int top = lua_gettop(L);
+	if (top < 1 || lua_type(L, top) != LUA_TSTRING)
+		return luaL_error(L, "expected string");
+	const char *s = lua_tostring(L, top);
+	assert(s != NULL);
+	/* Push the document table. */
+	lua_newtable(L);
+	enum {
+		ELEM,
+		TAG,
+		START_TAG,
+		END_TAG,
+		SPACE_AFTER_END_TAG,
+		ATTR,
+		ATTR_NAME,
+		ATTR_VALUE_SEP,
+		ATTR_VALUE_BEGIN,
+		ATTR_VALUE,
+		ATTR_VALUE_END,
+		ELEM_END,
+		DOC_END,
+	} state = ELEM;
+	const char *tag_name = NULL;
+	const char *attr_name = NULL;
+	const char *attr_value = NULL;
+	/*
+	 * Position of the current character in the input string recorded
+	 * as line and column numbers. Used for error reporting.
+	 */
+	int line = 1;
+	int column = 1;
+#define ERROR(msg) \
+	luaL_error(L, "XML decode error at line %d, column %d: %s", \
+		   line, column, (msg))
+	while (*s != '\0') {
+		/*
+		 * State machine:
+		 *  - 'break' skips the current character by leaving
+		 *    the switch-case.
+		 *  - 'continue' rechecks the state without skipping
+		 *    the current character.
+		 */
+		switch (state) {
+		/*
+		 * Expected an element. Currently, only tags are supported
+		 * (<section> or </section>) while values enclosed in tags are
+		 * considered invalid. Skip optional spaces and '<', then
+		 * transition to TAG.
+		 *
+		 * TODO: Handle values, such as <section>value</section>.
+		 */
+		case ELEM:
+			if (isspace(*s))
+				break; /* skip spaces */
+			if (*s != '<')
+				return ERROR("invalid token");
+			state = TAG;
+			break; /* skip '<' */
+		/*
+		 * If the current character is '/', skip it and transition to
+		 * END_TAG, otherwise transition to START_TAG.
+		 */
+		case TAG:
+			if (*s != '/') {
+				tag_name = s;
+				state = START_TAG;
+				continue;
+			}
+			if (lua_gettop(L) == top + 1)
+				return ERROR("invalid token");
+			tag_name = s + 1;
+			state = END_TAG;
+			break; /* skip '/' */
+		/*
+		 * Skip the start tag name, create a table for the tag
+		 * attributes and child elements, then transition to ATTR.
+		 */
+		case START_TAG: {
+			assert(tag_name != NULL);
+			if (isalnum(*s))
+				break; /* skip the tag name */
+			if ((!isspace(*s) && *s != '/' && *s != '>') ||
+			    s == tag_name)
+				return ERROR("invalid token");
+			/* Stack: parent table, ... */
+			assert(lua_gettop(L) >= top + 1);
+			assert(lua_istable(L, -1));
+			/* TODO: Check the tag name. */
+			lua_pushlstring(L, tag_name, s - tag_name);
```

Comment:
The TODO is to check beyond the `isalnum()` and emptiness checks above? If so, what check did you suppose here?

Reply:
The same question about `/* TODO: Check the attribute name. */` below.

Reply:
XML has naming rules, which are more complicated than that. I updated the comments to make it clear.

<details><summary>diff</summary>

```diff
diff --git a/src/lua/xml.c b/src/lua/xml.c
index fa4ad4035ff2..0eeba3500223 100644
--- a/src/lua/xml.c
+++ b/src/lua/xml.c
@@ -97,6 +97,7 @@ luaT_xml_decode(struct lua_State *L)
 		 */
 		case START_TAG: {
 			assert(tag_name != NULL);
+			/* TODO: Check names according to XML standard. */
 			if (isalnum(*s))
 				break; /* skip the tag name */
 			if ((!isspace(*s) && *s != '/' && *s != '>') ||
@@ -105,7 +106,6 @@ luaT_xml_decode(struct lua_State *L)
 			/* Stack: parent table, ... */
 			assert(lua_gettop(L) >= top + 1);
 			assert(lua_istable(L, -1));
-			/* TODO: Check the tag name. */
 			lua_pushlstring(L, tag_name, s - tag_name);
 			/*
 			 * Check that there's no attributes with the same name
@@ -198,6 +198,7 @@ luaT_xml_decode(struct lua_State *L)
 		 */
 		case ATTR_NAME:
 			assert(attr_name != NULL);
+			/* TODO: Check names according to XML standard. */
 			if (isalnum(*s))
 				break; /* skip the attribute name */
 			if ((!isspace(*s) && *s != '=') ||
@@ -205,7 +206,6 @@ luaT_xml_decode(struct lua_State *L)
 				return ERROR("invalid token");
 			/* Stack: attr table, tag name, parent table, ... */
 			assert(lua_gettop(L) >= top + 3);
-			/* TODO: Check the attribute name. */
 			lua_pushlstring(L, attr_name, s - attr_name);
 			/* Check that there's no duplicate attributes. */
 			lua_pushvalue(L, -1); /* attribute name */
diff --git a/src/lua/xml.h b/src/lua/xml.h
index d74ddba618cc..5e7ba706afa1 100644
--- a/src/lua/xml.h
+++ b/src/lua/xml.h
@@ -46,7 +46,10 @@ struct lua_State;
  * Limitations:
  *  - Element values, such as <section>value</section>, aren't supported.
  *  - Escape sequences in attribute values aren't supported.
- *  - Tag and attribute names aren't checked.
+ *  - Tag and attribute names aren't checked according to the XML standard.
+ *    The parser allows only digits and letters while in XML a name may also
+ *    contain dots, dashes, and underscores, and must start with a letter or
+ *    an underscore.
  */
 int
 luaT_xml_decode(struct lua_State *L);
```

</details>